### DISCUSSION TITLE
Android 14, Google Health not working

### DISCUSSION THREAD

**Discussion by daproc76** [NightscoutFoundation/xDrip#3130] [Bug reports]
I use xdrip+ on an Android 14 device. When I trie to connect xdrip+ with Google Health, I get the message app update required - connect the developer.

Thanks for this great App. I love it.
Best regards
daproc76

**Comment by corneliusroemer** [comment]
Can you share a screenshot? Which app says "App update required"? xdrip or Google health?

Is it just this modal that has been reported before?
- https://github.com/NightscoutFoundation/xDrip/discussions/3135
- https://github.com/NightscoutFoundation/xDrip/discussions/3114

If it's something else, more information would be useful as otherwise it's not very actionable.

This issue is part of the Android 14 collection: https://github.com/NightscoutFoundation/xDrip/discussions/3141

**Comment by daproc76** [comment]
![Screenshot_20231018-210453](https://github.com/NightscoutFoundation/xDrip/assets/146636815/0117dc56-837a-4382-b827-857dd16b3159)
When I switch Use Health Connect to on I get this Screen
![Screenshot_20231018-210439](https://github.com/NightscoutFoundation/xDrip/assets/146636815/9524c6c6-30b8-4eba-bec1-112643cf4230)

**Reply by corneliusroemer** [reply]
I can't reproduce on Android 14 with the latest nightly build. I don't get an error when I switch on Google Health connect.

What's your xdrip version/build?

**Comment by corneliusroemer** [comment]
I seem to have gotten further than you but now get this error: 
- https://github.com/NightscoutFoundation/xDrip/discussions/3143

**Comment by daproc76** [comment]
My Xdrip version is xDrip-plus-20231017-330f63b.apk this.
I use Android 14 on a Pixel 8 Pro

**Reply by corneliusroemer** [reply]
Interesting, I'm on c6b15c5-2023-10.08 on Pixel 8 Pro with Android 14 and I don't get your error. Notably, my version is _older_ than yours!

**Comment by daproc76** [comment]
![Screenshot_20231018-222112](https://github.com/NightscoutFoundation/xDrip/assets/146636815/043d52ff-0f6a-4107-bd5f-595b5166f634)

**Reply by corneliusroemer** [reply]
Hah, that's exactly the error I get too, see #3143

**Reply by corneliusroemer** [reply]
Do you know whether xdrip is supposed to be able to do anything with Google Health? is there documentation on the xdrip feature somewhere or did you just try it out because it was in the settings menu? this could be a feature under development, not yet working.

**Comment by daproc76** [comment]
![Screenshot_20231018-222401](https://github.com/NightscoutFoundation/xDrip/assets/146636815/4e81ec82-d287-469f-995d-6fb9fa8208b5)

**Comment by corneliusroemer** [comment]
I can reproduce your error! When I toggle the google health toggle off and back on, I now also get the "App update needed" screen that you reported here.

**Comment by daproc76** [comment]
On my old Mobile Oneplus 8 with Android 13 (Lineage 20) I can share my Blood sugar dates with Google Fit.

![Screenshot_20231018-223149](https://github.com/NightscoutFoundation/xDrip/assets/146636815/01c19fbc-c705-4800-92b3-8330222ea843)

**Reply by corneliusroemer** [reply]
Nice! How funny that you also just switched from a Oneplus 8 to Pixel 8 Pro - I just did the same!

**Reply by daproc76** [reply]
Yes that's really funny. Where are you from?

**Comment by daproc76** [comment]
![Screenshot_20231018-223355](https://github.com/NightscoutFoundation/xDrip/assets/146636815/903a0ae8-3a5d-45a5-9377-3893a6e7da53)

**Reply by corneliusroemer** [reply]
Yep that's the toast I got too

**Comment by iSaluki** [comment]
Yep, can confirm this issue on my Pixel 8.![Screenshot_20231019-173157.png](https://github.com/NightscoutFoundation/xDrip/assets/37915796/170c6823-b54d-4877-8dbd-6ad1ac2dce49)

**Comment by ah7675** [comment]
Android 14 has migrated HealthConnect from a user app to a system app.
https://developer.android.com/health-and-fitness/guides/health-connect/migrate/migrate-from-android-13-to-14

I have encountered the same error. From Pixel 4a to Pixel 8.

**Reply by databoy2k** [reply]
I set up a brand new Pixel 8 on Android 14. I ran into this issue in a few apps (the most irritating being Samsung's S-Health - didn't they develop Health Connect!?). It seems that the system migrated apps if you came from A13 and upgraded to 14, but there's no migration policy for old apps installed and asking to connect to Health Connect for the first time in A14.

From that page, is this the issue?
> In Android 14, you need to use an intent action, specified in the Jetpack SDK, which has different values based on the Android version it's acting on:
> 
> @get:JvmName("getHealthConnectSettingsAction") @JvmStatic val ACTION_HEALTH_CONNECT_SETTINGS

**Comment by Navid200** [comment]
@daproc76 Please click on "Create issue from discussion" in the right pane so that we can track this.

**Comment by rickykenny** [comment]
@Navid200 unsure if this comment belongs here in this discussion, or would you prefer it in the linked issue?

The migration guide outlines [changes in the Jetpack SDK](https://developer.android.com/health-and-fitness/guides/health-connect/migrate/migrate-from-android-13-to-14#permission-declaration). At a minimum, the permissions need to be declared differently. I'm unsure whether xDrip is affected by the other changes.

**Reply by Navid200** [reply]
A lot needs to happen for us to be able to update somethings.
If we only needed to support the latest phones, this would not be an issue.
But, xDrip is used worldwide by people who only have one thing in common, diabetes.
Therefore, being able to afford to get a phone with Android 11 or 12 or 13 or 14 may be a part of life for some, while for many it can be quite unreachable.

Google does not take that into consideration.  So, with every update Google makes it harder and harder for us to continue to support older versions.
We have retired Android 5.  
We would like to retire Android 6.  But, we have a lot to do.

So, in short, this may not be fixed overnight.  But, it will be fixed when possible.

**Reply by rickykenny** [reply]
I'm learning Android development now and I'm shocked by how fragmented the ecosystem is, I can only imagine how frustrating it is to maintain projects like xDrip. So a HUGE thanks to you all for your dedication 🙏🏽

Would you be open to a PR that addresses the Health Connect integration? This might be super naive, but it seems like it might be a good "first issue"? If so, I'll put together a proposal for review.

**Reply by Navid200** [reply]
It may be a bad idea to open a PR for something that another developer has been working on for a long time and is about to open a PR for.  
That's why developers can assign issues to themselves.  Then, a developer will know that another developer is working on something.

This is only come up after Android 14 was released.  I am not doing anything about it.
The lead developer is always working on moving all the different protocols forward peacefully.  

So, the worst thing that could happen is that your PR will not be merged.  If you are OK with that possibility, you can open a PR.

**Reply by rickykenny** [reply]
Ok thanks, I will give it a go and update you here if I manage to get a viable PR 🌞

**Reply by jamorham** [reply]
I'm actively working on this and may have an update fairly soon. Currently just testing it. The changes google have made are somewhat annoying and the product seems inconsistent on different platforms and developer unfriendly but such is life.

**Reply by Navid200** [reply]
@rickykenny 
Considering, it is best to be patient and wait on this.

**Comment by jamorham** [comment]
Please try the November 14th Nightly (available now) which should resolve this issue.

**Reply by rickykenny** [reply]
Amazing! Thanks so much, it is working for me 🙏🏽

**Comment by daproc76** [comment]
It works! THX

**Comment by iricigor** [comment]
I have a similar issue but in combination **Dexcom G6 app** + **Samsung Health**. Sync of vitals was working fine with Android 13, but as soon as I updated to A14, it stopped syncing. When I tried to reconfigure it, it silently failed. After uninstalling/reinstalling Dexcom and trying to configure it back, I was getting the same message "Dexcom app update required" from Samsung Health app. Not sure what are the underlaying connections between these apps, but it looks very similar.
